packages/gatsby-image/src/index.js

```diff
@@ -238,6 +238,18 @@ const listenToIntersections = (el, cb) => {
   }
 }
 
+const convertToCSSString = styleObject => {
+  if (!styleObject) return ""
+
+  const cssProps = Object.keys(styleObject)
+  const jsToCSSProp = capitalLetter => `-${capitalLetter[0].toLowerCase()}`
+
+  return cssProps.reduce((cssString, propName) => {
+    const propAsCSS = propName.replace(/([A-Z])/g, jsToCSSProp)
+    return cssString + `${propAsCSS}:${styleObject[propName]};`
+  }, "")
+}
+
```

Comment:
I wonder if there is any `react-dom` helper function we could use for it? It's for sure needed when we render react elements to html string, so would be nice to reuse it and not have to reimplement it.

Other option would be to actually use React to construct content of `<noscript>` instead of string concatenation and then we wouldn't need to make conversion like that ourselves. Then we could just set style to be an object 🤔

Reply:
Maybe one of those `react-dom` functions would do:
- https://github.com/facebook/react/blob/fe2cb525542443aaf1447c4069354c12659fd186/packages/react-dom/src/shared/CSSPropertyOperations.js#L24
- https://github.com/facebook/react/blob/fe2cb525542443aaf1447c4069354c12659fd186/packages/react-dom/src/server/ReactPartialRenderer.js#L220

Reply:
I don't think those functions are exported from `react-dom` module.

Reply:
yeah, I had a look and they don't seem to be exportet. I also had a search around the web to see if there was anything else/any libraries that are tested but had no luck.

Unfortunately the `noscript` tag seems to need a string as child (it calls `toString()` or similar on its children), so using React doesn't work either. It will just render as `[Object object]`.

Reply:
Update: I found the `ReactDOMServer.renderToString` function (https://reactjs.org/docs/react-dom-server.html) which seems to work.

Note that the current state of the code isn't cleaned up yet since I wanted to get a yes or no on my approach before making everything pretty

Reply:
the `ReactDOMServer` is problematic, because "server" part of `react-dom` will need to load in browser - functionally it probably will work, but it will also "bloat" our bundles.

The other approach could be figuring out why we use `dangerouslySetInnerHTML` in https://github.com/gatsbyjs/gatsby/blob/2ea2a8ec6e7665dcb655beda53ea1f87d4d26d81/packages/gatsby-image/src/index.js#L562-L574
instead of just treating that that regular component - right now we expect string there, and in master we manually construct it and this PR we actually use react to construct markup but there is still weird indirection there. Maybe we don't need to use it? Then we wouldn't need to look into using additional `ReactDOMServer` indirections?

Reply:
Unfortunately the `noscript` tag seems to convert all its children to string (as if it was) using the built-in `toString()` function, so just passing a component as a child doesn't work since it will just render as `[Object object]`. 

From what I could gather this seems to be the (currently) intended API of `noscript`: https://github.com/facebook/react/issues/15238

Reply:
Ah! Ok, this make sense right now. 

So I think the `ReactDOMServer` is only sensibly route to take, but what we need to make sure of is that we only include this for SSR (so .html is correctly generated) and that for client side it doesn't (as the `<noscript>` should have no effect in browser context anyway as you do need JS enabled in the first place to use browser bundle).

So curse of action I can see here could be something like that:
Instead of importing `react-dom/server` directly we have extra module (say `render-to-string.js`) that looks like this:

```js
if (process.env.BUILD_STAGE === `build-html` || process.env.BUILD_STAGE === `develop-html`) {
  const { renderToString } = require(`react-dom/server`)
  return renderToString
} else {
  module.exports = () => ``
}
```

This borrows from what we already have in `gatsby` core runtime ( https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/public-page-renderer.js )

Unfortunate reality of it is that we need to use CommonJS (require / module.exports) instead of ES modules (`import`/`export`), because ES modules can't handle conditional imports/exports. There are likely other ways to do similar things (by using webpack aliases, but `gatsby-image` is not gatsby plugin, so it doesn't have ability to modify gatsby's webpack configuration.

With setup like that we should be able to avoid adding `react-dom/server` to browser bundles.

What do you think about this approach?

Also /cc @wardpeet, @blainekasten - if anyone from the team would have objection about this, it might be you ;) so would love to hear your thoughts as well. Both about my proposition and maybe suggestions for alternatives

Reply:
@pieh How far away is the move to v3? 

This would be a stop gap solution to not use the string workaround(that this PR mostly removes via ReactDOMServer) only until then as we're only doing this until minimum React version can be bumped to >=16.6([Discussed a April 2019](https://github.com/gatsbyjs/gatsby/issues/13353)).

Is this worth going forward with until v3, or could we skip it?